File: DotNet Project V2.0/backend-dotnet/src/Emart.Domain/Purchase/ProductOffer.cs
namespace Emart.Domain.Purchase;

/// <summary>
/// Everything the PRODUCT PAGE (and the listing card) needs in order to render a product without
/// guessing anything. The backend returns all necessary information; the frontend just prints it.
/// </summary>
public sealed class ProductOffer
{
    public PurchaseMode Mode { get; }

    /// <summary>Cash a non-opted-in shopper pays: sale price, or MRP.</summary>
    public decimal SellingPrice { get; }

    public decimal MrpPrice { get; }

    /// <summary>Cash under the eMCard offer (modes 2 and 4); 0 otherwise.</summary>
    public decimal EmcardCashPrice { get; }

    /// <summary>Points under the eMCard offer (modes 3 and 4); 0 otherwise.</summary>
    public int PointsRequired { get; }

    /// <summary>False only for mode 3, where no cash is paid.</summary>
    public bool EarnsPoints { get; }

    public bool SaleActive { get; }

    public ProductOffer(
        PurchaseMode mode,
        decimal sellingPrice,
        decimal mrpPrice,
        decimal emcardCashPrice,
        int pointsRequired,
        bool earnsPoints,
        bool saleActive)
    {
        Mode = mode;
        SellingPrice = sellingPrice;
        MrpPrice = mrpPrice;
        EmcardCashPrice = emcardCashPrice;
        PointsRequired = Math.Max(0, pointsRequired);
        EarnsPoints = earnsPoints;
        SaleActive = saleActive;
    }

    /// <summary>Mode 2 only - the one mode the shopper can opt in and out of.</summary>
    public bool PointsOptional => Mode.IsOptional();

    /// <summary>Cash saved by taking the eMCard offer (modes 2 and 4).</summary>
    public decimal EmcardSavings
    {
        get
        {
            if (!Mode.RequiresCash() || EmcardCashPrice <= 0m || EmcardCashPrice >= SellingPrice)
            {
                return 0m;
            }

            return SellingPrice - EmcardCashPrice;
        }
    }
}

File: Java Project  V1.7/backend-spring/src/main/java/com/emart/service/purchase/ProductOffer.java
package com.emart.service.purchase;

import java.math.BigDecimal;

/**
 * Everything the PRODUCT PAGE (and the listing card) needs in order to
 * render a product without guessing anything.
 *
 * "The frontend should never guess the purchase mode. The backend must
 * return all necessary information through the API." - so this is what
 * the backend returns, and React just prints it:
 *
 *   CASH_ONLY           sellingPrice
 *   EMCARD_DISCOUNT     mrp, emcardCashPrice, savings, checkbox
 *   FULL_REDEMPTION     pointsRequired ("Redeem using EMCard Points")
 *   PARTIAL_REDEMPTION  emcardCashPrice + pointsRequired ("Cash + Points")
 */
public final class ProductOffer {

    private final PurchaseMode mode;
    private final BigDecimal sellingPrice;
    private final BigDecimal mrpPrice;
    private final BigDecimal emcardCashPrice;
    private final int pointsRequired;
    private final boolean earnsPoints;
    private final boolean saleActive;


    public ProductOffer(
            PurchaseMode mode,
            BigDecimal sellingPrice,
            BigDecimal mrpPrice,
            BigDecimal emcardCashPrice,
            int pointsRequired,
            boolean earnsPoints,
            boolean saleActive) {

        this.mode = mode;
        this.sellingPrice = nvl(sellingPrice);
        this.mrpPrice = nvl(mrpPrice);
        this.emcardCashPrice = nvl(emcardCashPrice);
        this.pointsRequired = Math.max(0, pointsRequired);
        this.earnsPoints = earnsPoints;
        this.saleActive = saleActive;
    }


    public PurchaseMode getMode() {
        return mode;
    }

    /** Cash a non-opted-in shopper pays: sale price, or MRP. */
    public BigDecimal getSellingPrice() {
        return sellingPrice;
    }

    public BigDecimal getMrpPrice() {
        return mrpPrice;
    }

    /** Cash under the eMCard offer (modes 2 and 4); 0 otherwise. */
    public BigDecimal getEmcardCashPrice() {
        return emcardCashPrice;
    }

    /** Points under the eMCard offer (modes 3 and 4); 0 otherwise. */
    public int getPointsRequired() {
        return pointsRequired;
    }

    /** False only for mode 3, where no cash is paid. */
    public boolean isEarnsPoints() {
        return earnsPoints;
    }

    public boolean isSaleActive() {
        return saleActive;
    }

    /** Mode 2 only - the one mode the shopper can opt in and out of. */
    public boolean isPointsOptional() {
        return mode != null && mode.isOptional();
    }

    /** Cash saved by taking the eMCard offer (modes 2 and 4). */
    public BigDecimal getEmcardSavings() {

        if (mode == null || !mode.requiresCash()
                || emcardCashPrice.signum() <= 0
                || emcardCashPrice.compareTo(sellingPrice) >= 0) {
            return BigDecimal.ZERO;
        }

        return sellingPrice.subtract(emcardCashPrice);
    }


    private static BigDecimal nvl(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value;
    }
}
